components/org.wso2.carbon.identity.application.authz.xacml/src/test/java/org.wso2.carbon.identity.application.authz.xacml/pip/AuthenticationContextAttributePIPTest.java

```diff
@@ -0,0 +1,265 @@
+        attributeValueList.add(attributeValue);
+        BagAttribute bagAttribute = spy(new BagAttribute(new URI(TYPE), attributeValueList));
+        mockStatic(FrameworkUtils.class);
+        AuthenticationContext authenticationContext = spy(new AuthenticationContext());
+        EvaluationResult context = spy(new EvaluationResult(attributeValue));
+        when(FrameworkUtils.getAuthenticationContextFromCache(anyString())).thenReturn(null);
+        when(evaluationCtx.getAttribute(any(URI.class), any(URI.class), anyString(), any(URI.class))).thenReturn(null);
+        assertEquals(authenticationContextAttributePIP.getAttributeValues(attributeType, attributeId, category,
+                issuer, evaluationCtx).size(), 0);
+
+        when(evaluationCtx.getAttribute(any(URI.class), any(URI.class), anyString(), any(URI.class)))
+                .thenReturn(context);
+        assertEquals(authenticationContextAttributePIP.getAttributeValues(attributeType, attributeId, category, issuer,
+                evaluationCtx).size(), 0);
+
+        when(context.getAttributeValue()).thenReturn(null);
+        assertEquals(authenticationContextAttributePIP.getAttributeValues(attributeType, attributeId, category, issuer,
+                evaluationCtx).size(), 0);
+
+        context = new EvaluationResult(bagAttribute);
+        when(evaluationCtx.getAttribute(any(URI.class), any(URI.class), anyString(), any(URI.class)))
+                .thenReturn(context);
+        when(bagAttribute.size()).thenReturn(0);
+        assertEquals(authenticationContextAttributePIP.getAttributeValues(attributeType, attributeId, category, issuer,
+                evaluationCtx).size(), 0);
+
+        when(bagAttribute.size()).thenCallRealMethod();
+        assertEquals(authenticationContextAttributePIP.getAttributeValues(attributeType, attributeId, category, issuer,
+                evaluationCtx).size(), 0);
+
+        when(FrameworkUtils.getAuthenticationContextFromCache(anyString())).thenReturn(authenticationContext);
+        category = new URI(XACMLAppAuthzConstants.AUTH_CONTEXT_PROPERTY_CATEGORY);
+        Map<String, String> propertyMap = new HashMap<>();
+        propertyMap.put("propertyKey", "propertyValue");
+        authenticationContext.setAuthenticatorProperties(propertyMap);
+        assertEquals(authenticationContextAttributePIP.getAttributeValues(attributeType, attributeId, category, issuer,
+                evaluationCtx).size(), 0);
+
+        category = new URI(XACMLAppAuthzConstants.AUTH_CONTEXT_REQ_PARAM_CATEGORY);
+        AuthenticationRequest authenticationRequest = new AuthenticationRequest();
+        Map<String, String[]> queryParams = new HashMap<>();
+        authenticationRequest.setRequestQueryParams(queryParams);
+        authenticationContext.setAuthenticationRequest(authenticationRequest);
+        assertEquals(authenticationContextAttributePIP.getAttributeValues(attributeType, attributeId, category, issuer,
```

Comment:
Looking at the asserts, looks to me like you can use a data provider here

Reply:
This test was written before we find a solution for PriviledgedCarbonContext. The problem is with the statement in line 128, the branching cannot be done with different inputs. And with deeper branching, it gets worse. I will try to refactor this as far as i could.

Reply:
There is also a known bug that data providers only works for primitives and basic objects.

Reply:
Yeah noticed the same. But there is a workaround for that too,
Basically, in the method that you get params from DataProvider we can set the type as Object and cast to specific type required

something like

@Test (dataprovider="fooProvider")
public foo(Object samlReponse, int i) {
      SAMLResponse response = (SAMLResponse) samlResponse;
}